src/Scout/ScoutEngine.php

```diff
@@ -0,0 +1,527 @@
+            $model = $model->first();
+        }
+
+        assert(in_array(Searchable::class, class_uses_recursive($model)), sprintf('Model "%s" must use "%s" trait', $model::class, Searchable::class));
+
+        return $this->mongodb->selectCollection($model->searchableAs());
+    }
+
+    /** Get the MongoDB collection used to index the provided model */
+    private function getIndexableCollection(Model|EloquentCollection $model): MongoDBCollection
+    {
+        if ($model instanceof EloquentCollection) {
+            $model = $model->first();
+        }
+
+        assert(in_array(Searchable::class, class_uses_recursive($model)), sprintf('Model "%s" must use "%s" trait', $model::class, Searchable::class));
+
+        return $this->mongodb->selectCollection($model->indexableAs());
+    }
+
+    private static function serialize(mixed $value): mixed
+    {
+        if ($value instanceof DateTimeInterface) {
+            return new UTCDateTime($value);
+        }
+
+        if ($value instanceof Serializable || ! is_iterable($value)) {
+            return $value;
+        }
+
+        // Convert Laravel Collections and other Iterators to arrays
+        if ($value instanceof Traversable) {
+            $value = iterator_to_array($value);
+        }
+
+        // Recursively serialize arrays
+        return array_map(self::serialize(...), $value);
+    }
+
+    private function usesSoftDelete(Model|EloquentCollection $model): bool
+    {
+        if ($model instanceof EloquentCollection) {
+            $model = $model->first();
+        }
+
+        return in_array(SoftDeletes::class, class_uses_recursive($model));
+    }
+
+    private function performSearchIndexOperation(Closure $closure): void
+    {
+        try {
+            $closure();
+        } catch (ServerException $exception) {
+            // @todo add error codes
+            if (in_array($exception->getCode(), [])) {
+                throw new \RuntimeException('Failed to perform search index operation. MongoDB Atlas cluster M10+ is required', 0, $exception);
+            }
+        }
+    }
+
+    private function getMapping(Model $model): array
+    {
+        $mapping = self::DEFAULT_DEFINITION;
+
+        if (method_exists($model, 'searchableMapping')) {
+            $mapping = $model->searchableMapping();
+        }
+
+        if ($this->usesSoftDelete($model)) {
+            // This field is a boolean represented with the integers 0 and 1
+            $mapping['fields']['__soft_deleted'] ??= [
+                'type' => 'number',
+                'representation' => 'int64',
+                'indexDoubles' => false,
+            ];
+        }
+
+        return $mapping;
+    }
+
+    /**
+     * Wait for the callback to return true, use it for asynchronous
+     * Atlas Search index management operations.
+     */
+    private function wait(Closure $callback): void
+    {
+        $timeout = hrtime()[0] + self::WAIT_TIMEOUT_SEC;
```

Comment:
This is likely a very minor edge case, but `hrtime()` could be unsupported and return `false`:

- https://github.com/php/php-src/blob/php-8.4.1/ext/standard/basic_functions.stub.php#L2186
- https://github.com/php/php-src/blob/php-8.4.1/ext/standard/hrtime.c#L49
- https://github.com/php/php-src/blob/php-8.4.1/Zend/zend_hrtime.h#L57

But if static analysis doesn't care about this then perhaps you shouldn't either.